File: Sources/Domain/CodeSigning/Devices/Device.swift
import Foundation

public struct Device: Sendable, Equatable, Identifiable, Codable {
    public let id: String
    public let name: String
    public let udid: String
    public let deviceClass: DeviceClass
    public let platform: BundleIDPlatform
    public let status: DeviceStatus
    public let model: String?
    public let addedDate: Date?

    public init(
        id: String,
        name: String,
        udid: String,
        deviceClass: DeviceClass,
        platform: BundleIDPlatform,
        status: DeviceStatus = .enabled,
        model: String? = nil,
        addedDate: Date? = nil
    ) {
        self.id = id
        self.name = name
        self.udid = udid
        self.deviceClass = deviceClass
        self.platform = platform
        self.status = status
        self.model = model
        self.addedDate = addedDate
    }

    public var isEnabled: Bool { status == .enabled }
}

extension Device: Presentable {
    public static var tableHeaders: [String] {
        ["ID", "Name", "UDID", "Class", "Status"]
    }
    public var tableRow: [String] {
        [id, name, udid, deviceClass.rawValue, status.rawValue]
    }
}

extension Device: AffordanceProviding {
    public var affordances: [String: String] {
        ["listDevices": "asc devices list"]
    }
}

public enum DeviceClass: String, Sendable, Equatable, Codable, CaseIterable {
    case appleWatch = "APPLE_WATCH"
    case iPad = "IPAD"
    case iPhone = "IPHONE"
    case iPod = "IPOD"
    case appleTV = "APPLE_TV"
    case mac = "MAC"
    case appleVisionPro = "APPLE_VISION_PRO"
}

public enum DeviceStatus: String, Sendable, Equatable, Codable {
    case enabled = "ENABLED"
    case disabled = "DISABLED"
}

File: apps/asc-web/command-center/src/code-signing/Device.ts
import type { Affordances } from '../shared/types.ts';

export type DeviceStatus = 'ENABLED' | 'DISABLED';

export class Device {
  constructor(
    readonly id: string,
    readonly name: string,
    readonly udid: string,
    readonly deviceClass: string,
    readonly model: string,
    readonly status: DeviceStatus,
    readonly affordances: Affordances,
  ) {}

  // ── Semantic Booleans ──

  get isEnabled(): boolean {
    return this.status === 'ENABLED';
  }

  // ── Display ──

  get displayName(): string {
    return `${this.name} (${this.model})`;
  }

  // ── Factory ──

  static fromJSON(json: Record<string, unknown>): Device {
    return new Device(
      json.id as string,
      json.name as string,
      json.udid as string,
      json.deviceClass as string,
      json.model as string,
      json.status as DeviceStatus,
      (json.affordances as Affordances) ?? {},
    );
  }
}
